pkg/cloudprovider/aws/fleet/capacity.go

```diff
@@ -18,153 +18,53 @@ import (
 	"context"
 	"fmt"
 
-	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
 	"github.com/awslabs/karpenter/pkg/cloudprovider"
 )
 
 // Capacity cloud provider implementation using AWS Fleet.
 type Capacity struct {
-	spec                   *v1alpha1.ProvisionerSpec
-	launchTemplateProvider *LaunchTemplateProvider
-	subnetProvider         *SubnetProvider
-	nodeFactory            *NodeFactory
-	instanceProvider       *InstanceProvider
+	spec             *v1alpha1.ProvisionerSpec
+	nodeFactory      *NodeFactory
+	packing          cloudprovider.Packer
+	instanceProvider *InstanceProvider
+	vpc              *VPCProvider
 }
 
 // Create a set of nodes given the constraints.
-func (c *Capacity) Create(ctx context.Context, constraints *cloudprovider.CapacityConstraints) (cloudprovider.CapacityPacking, error) {
-	// 1. Compute Constraints
-	launchTemplate, err := c.launchTemplateProvider.Get(ctx, c.spec.Cluster)
-	if err != nil {
-		return nil, fmt.Errorf("getting launch template, %w", err)
-	}
-	instancePackings, err := c.instanceProvider.GetPackings(ctx, constraints.Pods, constraints.Overhead)
+func (c *Capacity) Create(ctx context.Context, constraints *cloudprovider.CapacityConstraints) (cloudprovider.Packings, error) {
+	// 1. Compute Packing given the constraints
+	instancePackings, err := c.packing.Get(ctx, constraints)
 	if err != nil {
 		return nil, fmt.Errorf("computing bin packing, %w", err)
 	}
-	zonalSubnets, err := c.getConstrainedZonalSubnets(ctx, constraints)
-	if err != nil {
-		return nil, fmt.Errorf("getting zonal subnets, %w", err)
-	}
 
 	// 2. Create Instances
 	var instanceIds []*string
-	for _, instancePacking := range instancePackings {
-		instanceId, err := c.instanceProvider.Create(ctx, launchTemplate, instancePacking.InstanceTypeOptions, zonalSubnets)
+	for tempID, instancePacking := range instancePackings {
+		ec2InstanceID, err := c.instanceProvider.Create(ctx, instancePacking.InstanceTypeOptions, constraints, c.spec)
 		if err != nil {
 			// TODO Aggregate errors and continue
 			return nil, fmt.Errorf("creating capacity %w", err)
 		}
-		instanceIds = append(instanceIds, instanceId)
+		instancePackings[*ec2InstanceID] = instancePacking
+		delete(instancePackings, tempID)
```

Comment:
Whenever I see statements like "delete", "break", "continue", my overly-complex code detector starts going off. In this case, you're doing an in place mutation of a map and removing all of the dummy keys and replacing them with real ones. This is pretty tricky, especially if it ever needs to change. For example, it means that anyone solving "TODO Aggregate errors and continue" now needs to be really careful about making sure all of the right keys are added and delete when they change their logic.

A simpler "immutable" approach would simply construct a new map and only add the new packings to it. The old map will be garbage collected. However, given previous comments about where or not we even need "id", I think this code might change anyways.